### PR TITLE
Split out arrow-arith (#2594)

### LINKED ISSUE #2594: Split up Arrow Crate
**TLDR rather than fighting entropy lets just brute-force compilation**

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The arrow crate is getting rather large, and is starting to show up as a non-trivial bottleneck when compiling code, see #2170. There have been some efforts to reduce the amount of generated code, see #1858, but this is going to be a perpetual losing battle against new feature additions.

I think there are a couple of problems currently:

1. Limited build parallelism, especially if [codegen-units](https://doc.rust-lang.org/stable/rustc/codegen-options/#codegen-units) is set low
2. Upstream crates have to "depend" on functionality they don't need, e.g. `parquet` depending on compute kernels
3. Minor changes force large amounts of recompilation, with incremental compilation only helping marginally
4. Codegen is rarely linear in complexity, consequently larger codegen units take longer than the same amount of code in smaller units

All these conspire to often result in an `arrow` shaped hole in compilation, where CPUs are left idle.

Some numbers from my local machine

* Release with default features: 232 seconds
* Release with default features without comparison kernels: 150 seconds
* Release with default features without compute kernels: 70 seconds
* Release without default features without compute kernels: 60 seconds

**The vast majority of the time all bar a single core is idle.**

**Describe the solution you'd like**

I would like to propose we split up the arrow crate, into a number of sub-crates that are then re-exported by the top-level `arrow` crate. Users can then choose to depend on the batteries included `arrow` crate, or more granular crates.

Initially I would propose the following split:

* arrow-csv: CSV reader support
* arrow-ipc: IPC support
* arrow-json: JSON support (related to #2300)
* arrow-compute: contents of compute module
* arrow-test: arrow test_utils (not published)
* arrow-core: everything else

There is definitely scope for splitting up the crates further after this, in particular the comparison kernels might be a good candidate to live on their own, but I think lets start small and go from there. I suspect there is a fair amount of disentangling that will be necessary to achieve this.

**Describe alternatives you've considered**

Feature flags are another way this can be handled, however, they have a couple of limitations:

* It is impractical to test the full combinatorial explosion of combinations, which allows for bugs to sneak through
* They are unified for a target which limits build parallelism, just because say DataFusion depends on arrow with CSV support, shouldn't force the `parquet` crate to wait for this to compile before it can start compiling
* Poor UX:
  * Discoverability is limited, it can be hard to determine what features gate what functionality
  * Hard to determine if the feature flag set is minimal, no equivalent of cargo-udeps
  * It can be a non-trivial detective exercise to determine why a given feature is being enabled
  * Necessitate counter-intuitive hacks to play nicely in multi-crate workspaces - see [workspace hack](https://docs.rs/cargo-hakari/latest/cargo_hakari/about/index.html#what-are-workspace-hack-crates)

**Additional context**

@Jimexist recently drove an initiative to do something similar to DataFusion which has worked very well - https://github.com/apache/arrow-datafusion/issues/1750

FYI @alamb @jhorstmann @nevi-me

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2594

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-arith/src/temporal.rs [hunk 1]
```diff
@@ -17,18 +17,19 @@
 
 //! Defines temporal kernels for time and date related functions.
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:18]: These aren't technically arithmetic kernels, but they are similar in spirit, and putting them in arrow-cast would bring in a load of additional stuff

File: arrow-arith/src/arithmetic.rs [hunk 2]
```diff
@@ -22,29 +22,12 @@
...
-use crate::buffer::MutableBuffer;
-use crate::compute::kernels::arity::unary;
-use crate::compute::{
-    binary, binary_opt, try_binary, try_unary, try_unary_dyn, unary_dyn,
-};
-use crate::datatypes::{
-    ArrowNativeTypeOp, ArrowNumericType, DataType, Date32Type, Date64Type,
-    IntervalDayTimeType, IntervalMonthDayNanoType, IntervalUnit, IntervalYearMonthType,
-};
-#[cfg(feature = "dyn_arith_dict")]
```
Comment (alamb) [inline arrow-arith/src/arithmetic.rs:36]: what changed about the feature guards that allows for this simplification?

File: arrow-arith/src/arithmetic.rs [hunk 3]
```diff
@@ -332,11 +316,7 @@ where
 
             valid_chunks
                 .iter()
-                .zip(
-                    result_chunks
-                        .borrow_mut()
-                        .zip(left_chunks.borrow_mut().zip(right_chunks.borrow_mut())),
-                )
+                .zip((&mut result_chunks).zip((&mut left_chunks).zip(&mut right_chunks)))
```
Comment (alamb) [inline arrow-arith/src/arithmetic.rs:319]: this doesn't need mut references, is that the difference?

File: arrow/src/compute/kernels/mod.rs [hunk 4]
```diff
@@ -17,14 +17,9 @@
 
 //! Computation kernels on Arrow Arrays
 
-pub mod aggregate;
-pub mod arithmetic;
-pub mod arity;
-pub mod bitwise;
-pub mod boolean;
 pub mod limit;
```
Comment (alamb) [inline arrow/src/compute/kernels/mod.rs:20]: what is the deal with limit? Is it going away too?

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold 

cc @viirya

File: arrow/src/compute/kernels/mod.rs (hunk 4 above)
Comment (tustvold) [reply arrow/src/compute/kernels/mod.rs:20]: It is deprecated and can probably be removed at some point - it just calls through to Array::slice

File: arrow-arith/src/arithmetic.rs (hunk 2 above)
Comment (tustvold) [reply arrow-arith/src/arithmetic.rs:36]: Combination of wildcard imports and explicitly scoping a couple of types

File: arrow-arith/src/arithmetic.rs (hunk 3 above)
Comment (tustvold) [reply arrow-arith/src/arithmetic.rs:319]: It avoids needing BorrowMut in scope :sweat_smile:

File: arrow-arith/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// under the License.
+
+//! Arrow arithmetic and aggregation kernels
+
+pub mod aggregate;
+pub mod arithmetic;
+pub mod arity;
+pub mod bitwise;
+pub mod boolean;
+pub mod temporal;
```
Comment (viirya) [inline arrow-arith/src/lib.rs:25]: Maybe we can have arrow-misc crate to hold such kernels which are not exactly arithmetic or comparison kernels? I'm fine to put them in `arrow-arith` though.

**Review by viirya — COMMENTED** [review]
Looks good to me. 👍

**Pull request merged**